arangod/V8Server/v8-voccursor.cpp

```diff
@@ -157,6 +164,386 @@ static void JS_JsonCursor(v8::FunctionCallbackInfo<v8::Value> const& args) {
+    }
+    
+    if (args.Length() < 1 || args.Length() > 3) {
+      TRI_V8_THROW_EXCEPTION_USAGE("ArangoQueryStreamCursor(<queryString>, <bindVars>, <options>)");
+    }
+    
+    // get the query string
+    if (!args[0]->IsString()) {
+      TRI_V8_THROW_TYPE_ERROR("expecting string for <queryString>");
+    }
+    std::string const queryString(TRI_ObjectToString(args[0]));
+
+    // bind parameters
+    std::shared_ptr<VPackBuilder> bindVars;
+    
+    if (args.Length() > 1) {
+      if (!args[1]->IsUndefined() && !args[1]->IsNull() && !args[1]->IsObject()) {
+        TRI_V8_THROW_TYPE_ERROR("expecting object for <bindVars>");
+      }
+      if (args[1]->IsObject()) {
+        bindVars.reset(new VPackBuilder);
+        int res = TRI_V8ToVPack(isolate, *(bindVars.get()), args[1], false);
+        
+        if (res != TRI_ERROR_NO_ERROR) {
+          TRI_V8_THROW_EXCEPTION(res);
+        }
+      }
+    }
+    
+    // options
+    auto options = std::make_shared<VPackBuilder>();
+    if (args.Length() > 2) {
+      // we have options! yikes!
+      if (!args[2]->IsObject()) {
+        TRI_V8_THROW_TYPE_ERROR("expecting object for <options>");
+      }
+      
+      int res = TRI_V8ToVPack(isolate, *options, args[2], false);
+      if (res != TRI_ERROR_NO_ERROR) {
+        TRI_V8_THROW_EXCEPTION(res);
+      }
+    } else {
+      VPackObjectBuilder guard(options.get());
+    }
+    size_t batchSize = VelocyPackHelper::getNumericValue<size_t>(options->slice(),
+                                                                 "batchSize", 1000);
+    
+    const bool contextOwnedByExterior = transaction::V8Context::isEmbedded();
+    TRI_vocbase_t* vocbase = v8g->_vocbase;
+    TRI_ASSERT(vocbase != nullptr);
+    auto* cursors = vocbase->cursorRepository(); // create a cursor
+    double ttl = std::numeric_limits<double>::max();
+    // specify ID 0 so it uses the external V8 context
+    auto cc = cursors->createQueryStream(queryString,
+                                         std::move(bindVars),
+                                         std::move(options),
+                                         batchSize, ttl,
+                                         contextOwnedByExterior);
+    TRI_DEFER(cc->release());
+    // args.Holder() is supposedly better than args.This()
+    auto self = new V8Cursor(isolate, args.Holder(), *vocbase, cc->id());
+    Result r = self->fetchData(cc);
+    if (r.fail()) {
+      TRI_V8_THROW_EXCEPTION(r);
+    }
+    // do not delete self, its owned by V8 now
+    
+    TRI_V8_RETURN(args.This());
+    TRI_V8_TRY_CATCH_END
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////
+  /// @brief ArangoQueryStreamCursor.prototype.toArray = ...
+  ////////////////////////////////////////////////////////////////////////////////
+
+  static void toArray(v8::FunctionCallbackInfo<v8::Value> const& args) {
+    v8::Isolate* isolate = args.GetIsolate();
```

Comment:
The code in this function should be wrapped into some `TRI_V8_TRY_CATCH_BEGIN` ... `TRI_V8_TRY_CATCH_END`.